[GLib] Replace GtkSettings usage with port agnostic settings

TingPing opened the pull request:
#### 41fb212daa290d014729a7005b3de77f30454657
<pre>
[GLib] Replace GtkSettings usage with port agnostic settings
<a href="https://bugs.webkit.org/show_bug.cgi?id=278875">https://bugs.webkit.org/show_bug.cgi?id=278875</a>

Reviewed by Carlos Garcia Campos.

This exposes settings as a singleton in WebCore that both
GTK and WPE can use.

For the GTK port it still reads GtkSettings to set all
of the values.

WPE doesn&apos;t expose an API to set these yet and just
returns default values.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/platform/SourcesGLib.txt:
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
(WebCore::ScrollbarThemeAdwaita::usesOverlayScrollbars const):
(WebCore::ScrollbarThemeAdwaita::handleMousePressEvent):
* Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
(WebCore::ThemeAdwaita::ThemeAdwaita):
(WebCore::ThemeAdwaita::refreshSettings):
(WebCore::ThemeAdwaita::refreshGtkSettings): Deleted.
* Source/WebCore/platform/adwaita/ThemeAdwaita.h:
* Source/WebCore/platform/glib/SystemSettings.cpp: Added.
(WebCore::SystemSettings::singleton):
(WebCore::SystemSettings::updateSettings):
(WebCore::SystemSettings::hintStyle const):
(WebCore::SystemSettings::subpixelOrder const):
(WebCore::SystemSettings::antialiasMode const):
(WebCore::SystemSettings::addObserver):
(WebCore::SystemSettings::removeObserver):
* Source/WebCore/platform/glib/SystemSettings.h: Copied from Source/WebKit/Shared/gtk/GtkSettingsState.h.
(WebCore::SystemSettings::settingsState const):
(WebCore::SystemSettings::themeName const):
(WebCore::SystemSettings::darkMode const):
(WebCore::SystemSettings::fontName const):
(WebCore::SystemSettings::cursorBlink const):
(WebCore::SystemSettings::cursorBlinkTime const):
(WebCore::SystemSettings::xftDPI const):
(WebCore::SystemSettings::overlayScrolling const):
(WebCore::SystemSettings::primaryButtonWarpsSlider const):
(WebCore::SystemSettings::enableAnimations const):
* Source/WebCore/platform/graphics/gtk/SystemFontDatabaseGTK.cpp:
(WebCore::SystemFontDatabase::platformSystemFontShorthandInfo):
* Source/WebCore/platform/gtk/GtkUtilities.cpp:
(WebCore::defaultGtkSystemFont):
(WebCore::shouldUseOverlayScrollbars):
* Source/WebCore/platform/gtk/PlatformScreenGtk.cpp:
(WebCore::fontDPI):
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.cpp:
(WebCore::ScrollbarThemeGtk::ScrollbarThemeGtk):
(WebCore::ScrollbarThemeGtk::handleMousePressEvent):
(WebCore::themeChangedCallback): Deleted.
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::caretBlinkInterval const):
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.h:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/Shared/glib/SystemSettings.serialization.in: Copied from Source/WebKit/Shared/gtk/ArgumentCodersGtk.serialization.in.
* Source/WebKit/Shared/gtk/ArgumentCodersGtk.serialization.in:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::effectiveAppearanceIsDark const):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseDispose):
(webkitWebViewBaseConstructed):
(webkitWebViewBaseCreateWebPage):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/UIProcess/glib/SystemSettingsManager.cpp: Added.
(WebKit::SystemSettingsManager::themeName const):
(WebKit::SystemSettingsManager::darkMode const):
(WebKit::SystemSettingsManager::fontName const):
(WebKit::SystemSettingsManager::xftAntialias const):
(WebKit::SystemSettingsManager::xftHinting const):
(WebKit::SystemSettingsManager::xftHintStyle const):
(WebKit::SystemSettingsManager::xftRGBA const):
(WebKit::SystemSettingsManager::xftDPI const):
(WebKit::SystemSettingsManager::cursorBlink const):
(WebKit::SystemSettingsManager::cursorBlinkTime const):
(WebKit::SystemSettingsManager::primaryButtonWarpsSlider const):
(WebKit::SystemSettingsManager::overlayScrolling const):
(WebKit::SystemSettingsManager::enableAnimations const):
(WebKit::SystemSettingsManager::initialize):
(WebKit::SystemSettingsManager::settingsDidChange):
* Source/WebKit/UIProcess/glib/SystemSettingsManager.h: Renamed from Source/WebKit/UIProcess/gtk/GtkSettingsManager.h.
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/gtk/Display.cpp:
(WebKit::Display::Display):
* Source/WebKit/UIProcess/gtk/ScreenManagerGtk.cpp:
* Source/WebKit/UIProcess/gtk/SystemSettingsManagerGtk.cpp: Renamed from Source/WebKit/UIProcess/gtk/GtkSettingsManager.cpp.
(WebKit::SystemSettingsManager::themeName const):
(WebKit::SystemSettingsManager::darkMode const):
(WebKit::SystemSettingsManager::fontName const):
(WebKit::SystemSettingsManager::xftAntialias const):
(WebKit::SystemSettingsManager::xftHinting const):
(WebKit::SystemSettingsManager::xftHintStyle const):
(WebKit::SystemSettingsManager::xftRGBA const):
(WebKit::SystemSettingsManager::xftDPI const):
(WebKit::SystemSettingsManager::cursorBlink const):
(WebKit::SystemSettingsManager::cursorBlinkTime const):
(WebKit::SystemSettingsManager::primaryButtonWarpsSlider const):
(WebKit::SystemSettingsManager::overlayScrolling const):
(WebKit::SystemSettingsManager::enableAnimations const):
(WebKit::SystemSettingsManager::SystemSettingsManager):
* Source/WebKit/WebProcess/glib/SystemSettingsProxy.cpp: Renamed from Source/WebKit/Shared/gtk/GtkSettingsState.h.
(WebKit::SystemSettingsProxy::initialize):
(WebKit::SystemSettingsProxy::SystemSettingsProxy):
(WebKit::SystemSettingsProxy::didChange):
* Source/WebKit/WebProcess/glib/SystemSettingsProxy.h: Renamed from Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.h.
* Source/WebKit/WebProcess/glib/SystemSettingsProxy.messages.in: Renamed from Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.messages.in.
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::platformInitializeWebProcess):
(WebKit::WebProcess::platformTerminate):
* Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.cpp: Removed.

Canonical link: <a href="https://commits.webkit.org/284171@main">https://commits.webkit.org/284171@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf4499614ff2ef676d29f4e4dff4082b81858bb5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68557 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47949 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21216 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72626 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19702 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70674 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55745 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19518 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54694 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13100 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71624 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43823 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59193 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35146 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40490 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16616 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18059 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62443 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16963 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74320 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12528 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16217 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62203 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12568 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59272 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62175 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15235 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10120 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3734 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43750 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44824 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46018 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44566 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/PlatformGTK.cmake:
```diff
@@ -50,6 +50,8 @@ list(APPEND WebCore_PRIVATE_FRAMEWORK_HEADERS
 
     platform/glib/ApplicationGLib.h
+    platform/glib/SettingsStateGLib.h
+    platform/glib/SystemSettingsGLib.h
```
I know we don't always follow the pattern, but in general FooGLib means the GLib implementation of Foo. So, I usually prefer to use GLibFoo for a GLib specific class or even just Foo. I think SystemSettings would work fine here.

carlosgcampos commented on Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
```diff
@@ -77,6 +80,8 @@ bool ScrollbarThemeAdwaita::usesOverlayScrollbars() const
 {
 #if PLATFORM(GTK)
     return shouldUseOverlayScrollbars();
```
I think we can move the env var check here, and then just use the same code for GTK and WPE

carlosgcampos commented on Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
```diff
@@ -323,14 +328,8 @@ ScrollbarButtonPressAction ScrollbarThemeAdwaita::handleMousePressEvent(Scrollba
-            return gtkWarpsSlader;
-        }();
+#if PLATFORM(GTK) || PLATFORM(WPE)
+        warpSlider = SystemSettingsGLib::singleton().primaryButtonWarpsSlider().value_or(true);
```
Why true by default? It was false before, no?

TingPing replied:
It was never false by default, it was just misleading code.

`g_object_get()` always overwrites the value passed in, and in GTK the default value was TRUE.

carlosgcampos commented on Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
```diff
@@ -51,46 +49,34 @@ Theme& Theme::singleton()
-        m_prefersContrast = !g_strcmp0(gtkThemeName.get(), "HighContrast") || !g_strcmp0(gtkThemeName.get(), "HighContrastInverse");
+    if (auto themeName = SystemSettingsGLib::singleton().themeName())
+        m_prefersContrast = themeName == "HighContrast"_s || themeName == "HighContrastInverse"_s;
 #endif // !USE(GTK4)
```
If this is specific to GTK3 and this method is also called for WPE now, we should probably use `#if PLATFORM(GTK) && !USE(GTK4)`

carlosgcampos commented on Source/WebCore/platform/glib/SettingsStateGLib.h:
```diff
@@ -25,16 +26,12 @@
-namespace WebKit {
-
-struct GtkSettingsState {
+struct SettingsStateGLib {
```
SettingsState is confusing, since we have Settings class, I would use SystemSettingsState. I think this could be moved to SystemSettings.h to make it even clearer that this belongs to SystemSettings.

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettingsGLib.cpp:
```diff
@@ -0,0 +1,165 @@
+
+#include "config.h"
+#include "SystemSettingsGLib.h"
+
```
#if PLATFORM(GTK) || PLATFORM(WPE)

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettingsGLib.cpp:
```diff
@@ -0,0 +1,165 @@
+
+SystemSettingsGLib::SystemSettingsGLib()
+{
+}
```
```suggestion
SystemSettingsGLib::SystemSettingsGLib() = default;
```

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettingsGLib.cpp:
```diff
@@ -0,0 +1,165 @@
+void SystemSettingsGLib::applySettings(SettingsStateGLib&& state)
+{
+    bool shouldUpdateStyle = false;
+    if (m_settings.themeName != state.themeName) {
```
The state members are optional, so if the member is nulltopt it means the value hasn't changed. So, if we have a theme name already and it hasn't changed, this condition will be true, but we don't want to change the setting at all. I don't think we need to compare the values, if it has changed, the value should always be different.

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettingsGLib.h:
```diff
@@ -0,0 +1,68 @@
+class SystemSettingsGLib {
+    WTF_MAKE_NONCOPYABLE(SystemSettingsGLib);
+    friend NeverDestroyed<SystemSettingsGLib>;
+public:
```
Ideally, the state class could be a nested class here (SystemSettings::State), but since we need it for IPC serialization, we could define here something like `using State = SystemSettingsState` and then we can use SystemSettings::State everywhere except for IPC messages.

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettingsGLib.h:
```diff
@@ -0,0 +1,68 @@
+    static SystemSettingsGLib& singleton();
+
+    void applySettings(SettingsStateGLib&&);
+    const SettingsStateGLib& settingsState() { return m_settings; }
```
```suggestion
    const SettingsStateGLib& settingsState() const { return m_settings; }
```

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettingsGLib.h:
```diff
@@ -0,0 +1,68 @@
+    void applyAntialiasSettings();
+
+    SettingsStateGLib m_settings;
+    HashMap<void*, Function<void(const WebCore::SettingsStateGLib&)>> m_observers;
```
```suggestion
    HashMap<void*, Function<void(const SettingsStateGLib&)>> m_observers;
```

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettingsGLib.h:
```diff
@@ -0,0 +1,68 @@
+    void applyHintingSettings();
+    void applyAntialiasSettings();
+
+    SettingsStateGLib m_settings;
```
```suggestion
    SettingsStateGLib m_state;
```

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettingsGLib.h:
```diff
@@ -0,0 +1,68 @@
+ */
+
+#pragma once
+
```
#if PLATFORM(GTK) || PLATFORM(WPE)

carlosgcampos commented on Source/WebCore/platform/gtk/GtkUtilities.cpp:
```diff
@@ -141,13 +142,16 @@ WallTime wallTimeForEvent(const GdkEvent* event)
-    g_object_get(gtk_settings_get_default(), "gtk-font-name", &fontString.outPtr(), nullptr);
+    auto fontName = SystemSettingsGLib::singleton().fontName();
+    if (!fontName || fontName->isEmpty())
+        return "Sans"_s;
```
Why Sans? We were not handling this case before, I guess gtk never returns null or empty string for the setting?

TingPing replied:
GTKs default is `Sans 10`.

carlosgcampos commented on Source/WebCore/platform/gtk/GtkUtilities.cpp:
```diff
@@ -141,13 +142,16 @@ WallTime wallTimeForEvent(const GdkEvent* event)
+    if (auto* spaceChar = strrchr(fontString, ' '))
         *spaceChar = '\0';
-    return String::fromUTF8(fontString.get());
+    return String::fromUTF8(fontString);
```
You can do this with WTFString API now and avoid utf8 back and froth conversions

carlosgcampos commented on Source/WebCore/platform/gtk/ScrollbarThemeGtk.cpp:
```diff
@@ -46,19 +45,19 @@ ScrollbarTheme& ScrollbarTheme::nativeTheme()
+        SystemSettingsGLib::singleton().addObserver([](const SettingsStateGLib& state) mutable {
+            if (state.themeName || state.overlayScrolling)
+                ScrollbarTheme::theme().themeChanged();
+        }, this);
```
Since this is a singleton, I guess this could be notified from the SystemSettings like we do with RenderTheme

carlosgcampos commented on Source/WebKit/PlatformGTK.cmake:
```diff
@@ -63,14 +63,15 @@ list(APPEND WebKit_MESSAGES_IN_FILES
 endif ()
 
 list(APPEND WebKit_SERIALIZATION_IN_FILES
+    Shared/glib/ArgumentCodersGLib.serialization.in
```
This is not a generic file for glib coders, it should be called SystemSettingsState.serialization.in

carlosgcampos commented on Source/WebKit/Shared/glib/ArgumentCodersGLib.serialization.in:
```diff
@@ -0,0 +1,36 @@
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+struct WebCore::SettingsStateGLib {
```
If we move the struct definition to SystemSettings.h here we need to add:

```
header: <WebCore/SystemSettings.h>
```

And probably  [CustomHeader] too before the struct

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -853,7 +854,7 @@ static void webkitWebViewBaseDispose(GObject* gobject)
 #endif
 
-    GtkSettingsManager::singleton().removeObserver(webView);
+    SystemSettingsGLib::singleton().removeObserver(webView);
```
I don't think it's a good idea to use the SystremSettings singleton from the UI process, since it's notifying about changes to RenderTheme, for example, that only makes sense in the web process. Here can use SystemSettings::State as the store for the UI process manager.

TingPing replied:
How should each view get notified of the settings change?

Maybe we could just have a separate `applySettings` function only for the UI process that doesn't update external state like `RenderTheme`.

TingPing replied:
Really I think it would be best if it didn’t call out anywhere rather using observers everywhere needed, so it’s always safe.

carlosgcampos replied:
I agree, it would be better to have the system settings object not aware of other webcore classes, and use observers everywhere.

carlosgcampos commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -163,8 +163,9 @@ void WebProcessPool::platformInitializeWebProcess(const WebProcessProxy& process
         parameters.accessibilityBusAddress = m_sandboxEnabled && shouldUseBubblewrap() ? sandboxedAccessibilityBusAddress() : accessibilityBusAddress();
 #endif
 
+    parameters.systemSettings = WebCore::SystemSettingsGLib::singleton().settingsState();
```
We probably need a SystemSettingsManager or SystemSettingsProvider or something like that with implementations for GTK and WPE to be used in the UI process.

carlosgcampos commented on Source/WebKit/UIProcess/gtk/GtkSettingsManager.cpp:
```diff
@@ -134,61 +133,62 @@ bool GtkSettingsManager::enableAnimations() const
 
     for (auto& processPool : WebProcessPool::allProcessPools())
-        processPool->sendToAllProcesses(Messages::GtkSettingsManagerProxy::SettingsDidChange(state));
+        processPool->sendToAllProcesses(Messages::SystemSettingsProxyGLib::SettingsDidChange(changedState));
```
```suggestion
        processPool->sendToAllProcesses(Messages::SystemSettingsProxyGLib::DidChange(changedState));
```

mcatanzaro commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -173,7 +173,7 @@ void InlineItemsBuilder::computeInlineBoxBoundaryTextSpacingsIfNeeded(const Inli
 
         size_t boundaryDepth = std::min(currentCharacterDepth, lastCharacterDepth);
-        ASSERT((inlineBoxStartIndexesOnInlineItemsList.size() - 1 - (currentCharacterDepth - boundaryDepth)) >= 0);
+        // ASSERT((inlineBoxStartIndexesOnInlineItemsList.size() - 1 - (currentCharacterDepth - boundaryDepth)) >= 0);
```
!

carlosgcampos commented on Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
```diff
@@ -75,8 +78,10 @@ void ScrollbarThemeAdwaita::updateScrollbarOverlayStyle(Scrollbar& scrollbar)
-    return shouldUseOverlayScrollbars();
+#if PLATFORM(GTK) || PLATFORM(WPE)
+    if (!g_strcmp0 (g_getenv ("GTK_OVERLAY_SCROLLING"), "0"))
+        return false;
```
This should be in a GTK only ifdef and only for GTK3.

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettings.cpp:
```diff
@@ -0,0 +1,176 @@
+{
+    bool shouldUpdateStyle = false;
+    bool shouldUpdateScrollbarTheme = false;
+    if (m_state.themeName) {
```
```suggestion
    if (state.themeName) {
```

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettings.cpp:
```diff
@@ -0,0 +1,176 @@
+        m_state.fontName = WTFMove(state.fontName);
+
+    bool hintingSettingsDidChange = false;
+    if (m_state.xftHinting) {
```
```suggestion
    if (state.xftHinting) {
```

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettings.cpp:
```diff
@@ -0,0 +1,176 @@
+        hintingSettingsDidChange = true;
+    }
+
+    if (m_state.xftHintStyle) {
```
```suggestion
    if (state.xftHintStyle) {
```

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettings.cpp:
```diff
@@ -0,0 +1,176 @@
+    }
+
+    bool antialiasSettingsDidChange = false;
+    if (m_state.xftAntialias) {
```
```suggestion
    if (state.xftAntialias) {
```

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettings.cpp:
```diff
@@ -0,0 +1,176 @@
+        antialiasSettingsDidChange = true;
+    }
+
+    if (m_state.xftRGBA) {
```
```suggestion
    if (state.xftRGBA) {
```

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettings.cpp:
```diff
@@ -0,0 +1,176 @@
+        ScrollbarTheme::theme().themeChanged();
+
+    if (shouldUpdateStyle)
+        Page::updateStyleForAllPagesAfterGlobalChangeInEnvironment();
```
I think we could indeed have a single observer in the web process to update style, font settings, theme, etc. But then we should probably rename applySettings to updateSettings

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettings.h:
```diff
@@ -0,0 +1,88 @@
+    std::optional<int> xftDPI() const { return m_state.xftDPI; }
+    std::optional<bool> overlayScrolling() const { return m_state.overlayScrolling; };
+    std::optional<bool> primaryButtonWarpsSlider() const { return m_state.primaryButtonWarpsSlider; };
+    std::optional<bool> enableAnimations() const { return m_state.enableAnimations; };
```
```suggestion
    std::optional<String> themeName() const { return m_state.themeName; }
    std::optional<String> fontName() const { return m_state.fontName; }
    std::optional<bool> cursorBlink() const { return m_state.cursorBlink; }
    std::optional<int> cursorBlinkTime() const { return m_state.cursorBlinkTime; }
    std::optional<int> xftDPI() const { return m_state.xftDPI; }
    std::optional<bool> overlayScrolling() const { return m_state.overlayScrolling; }
    std::optional<bool> primaryButtonWarpsSlider() const { return m_state.primaryButtonWarpsSlider; }
    std::optional<bool> enableAnimations() const { return m_state.enableAnimations; }
```

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -2463,12 +2463,11 @@ static void webkitWebViewBaseConstructed(GObject* object)
+    //     if (state.xftDPI)
+    //         refreshInternalScaling(webViewBase);
+    //     webViewBase->priv->pageProxy->effectiveAppearanceDidChange();
+    // }, viewWidget);
```
This needs to be implemented again

carlosgcampos commented on Source/WebKit/UIProcess/glib/SystemSettingsManager.cpp:
```diff
@@ -0,0 +1,165 @@
+    for (auto& processPool : WebProcessPool::allProcessPools())
+        processPool->sendToAllProcesses(Messages::SystemSettingsProxy::DidChange(changedState));
+
+    // SystemSettings::singleton().applySettings(WTFMove(changedState));
```
Is this actually a wip patch or draft?

TingPing replied:
I just left in the part I had an open discussion over.

carlosgcampos commented on Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
```diff
@@ -75,8 +78,12 @@ void ScrollbarThemeAdwaita::updateScrollbarOverlayStyle(Scrollbar& scrollbar)
-#if PLATFORM(GTK)
-    return shouldUseOverlayScrollbars();
+#if PLATFORM(GTK) && !USE(GTK4)
+    if (!g_strcmp0 (g_getenv ("GTK_OVERLAY_SCROLLING"), "0"))
```
```suggestion
    if (!g_strcmp0(g_getenv("GTK_OVERLAY_SCROLLING"), "0"))
```

carlosgcampos commented on Source/WebCore/platform/glib/SystemSettings.cpp:
```diff
@@ -0,0 +1,191 @@
+    if (state.enableAnimations)
+        m_state.enableAnimations = state.enableAnimations;
+
+    if (applyChanges) {
```
I thought the plan was to use observers. We can add an observer to the web process to apply these settings, no?

carlosgcampos commented on Source/WebKit/UIProcess/glib/SystemSettingsManager.cpp:
```diff
@@ -0,0 +1,165 @@
+
+#endif // !PLATFORM(GTK)
+
+void SystemSettingsManager::initialize()
+{
+    static NeverDestroyed<SystemSettingsManager> manager;
+}
```
I don't understand this

TingPing replied:
The manager has to be created after GtkSettings is but before anything uses it, so we have a method to create it. This function just creates it on first call.

Previously the manager just needed to exist before the webprocess launched as the UI process didn't rely on it.

carlosgcampos replied:
Is that the case being a static var? Don't you need to use LazyNeverDestroyed for that to work?

TingPing replied:
In testing it definitely worked but `LazyNeverDestroyed` sounds like a perfect fit so I'll switch to it, maybe it handles some edge case.

TingPing replied:
`LazyNeverDestroyed` lets you construct a static variable in a thread-safe way. A static `NeverDestroyed` is just immediately constructed when the function is called, but only once. So i think this is fine.

TingPing committed:
[GLib] Replace GtkSettings usage with port agnostic settings
https://bugs.webkit.org/show_bug.cgi?id=278875

Reviewed by Carlos Garcia Campos.

This exposes settings as a singleton in WebCore that both
GTK and WPE can use.

For the GTK port it still reads GtkSettings to set all
of the values.

WPE doesn't expose an API to set these yet and just
returns default values.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/platform/SourcesGLib.txt:
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
(WebCore::ScrollbarThemeAdwaita::usesOverlayScrollbars const):
(WebCore::ScrollbarThemeAdwaita::handleMousePressEvent):
* Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
(WebCore::ThemeAdwaita::ThemeAdwaita):
(WebCore::ThemeAdwaita::refreshSettings):
(WebCore::ThemeAdwaita::refreshGtkSettings): Deleted.
* Source/WebCore/platform/adwaita/ThemeAdwaita.h:
* Source/WebCore/platform/glib/SystemSettings.cpp: Added.
(WebCore::SystemSettings::singleton):
(WebCore::SystemSettings::updateSettings):
(WebCore::SystemSettings::hintStyle const):
(WebCore::SystemSettings::subpixelOrder const):
(WebCore::SystemSettings::antialiasMode const):
(WebCore::SystemSettings::addObserver):
(WebCore::SystemSettings::removeObserver):
* Source/WebCore/platform/glib/SystemSettings.h: Copied from Source/WebKit/Shared/gtk/GtkSettingsState.h.
(WebCore::SystemSettings::settingsState const):
(WebCore::SystemSettings::themeName const):
(WebCore::SystemSettings::darkMode const):
(WebCore::SystemSettings::fontName const):
(WebCore::SystemSettings::cursorBlink const):
(WebCore::SystemSettings::cursorBlinkTime const):
(WebCore::SystemSettings::xftDPI const):
(WebCore::SystemSettings::overlayScrolling const):
(WebCore::SystemSettings::primaryButtonWarpsSlider const):
(WebCore::SystemSettings::enableAnimations const):
* Source/WebCore/platform/graphics/gtk/SystemFontDatabaseGTK.cpp:
(WebCore::SystemFontDatabase::platformSystemFontShorthandInfo):
* Source/WebCore/platform/gtk/GtkUtilities.cpp:
(WebCore::defaultGtkSystemFont):
(WebCore::shouldUseOverlayScrollbars):
* Source/WebCore/platform/gtk/PlatformScreenGtk.cpp:
(WebCore::fontDPI):
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.cpp:
(WebCore::ScrollbarThemeGtk::ScrollbarThemeGtk):
(WebCore::ScrollbarThemeGtk::handleMousePressEvent):
(WebCore::themeChangedCallback): Deleted.
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::caretBlinkInterval const):
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.h:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/Shared/glib/SystemSettings.serialization.in: Copied from Source/WebKit/Shared/gtk/ArgumentCodersGtk.serialization.in.
* Source/WebKit/Shared/gtk/ArgumentCodersGtk.serialization.in:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::effectiveAppearanceIsDark const):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseDispose):
(webkitWebViewBaseConstructed):
(webkitWebViewBaseCreateWebPage):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/UIProcess/glib/SystemSettingsManager.cpp: Added.
(WebKit::SystemSettingsManager::themeName const):
(WebKit::SystemSettingsManager::darkMode const):
(WebKit::SystemSettingsManager::fontName const):
(WebKit::SystemSettingsManager::xftAntialias const):
(WebKit::SystemSettingsManager::xftHinting const):
(WebKit::SystemSettingsManager::xftHintStyle const):
(WebKit::SystemSettingsManager::xftRGBA const):
(WebKit::SystemSettingsManager::xftDPI const):
(WebKit::SystemSettingsManager::cursorBlink const):
(WebKit::SystemSettingsManager::cursorBlinkTime const):
(WebKit::SystemSettingsManager::primaryButtonWarpsSlider const):
(WebKit::SystemSettingsManager::overlayScrolling const):
(WebKit::SystemSettingsManager::enableAnimations const):
(WebKit::SystemSettingsManager::initialize):
(WebKit::SystemSettingsManager::settingsDidChange):
* Source/WebKit/UIProcess/glib/SystemSettingsManager.h: Renamed from Source/WebKit/UIProcess/gtk/GtkSettingsManager.h.
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/gtk/Display.cpp:
(WebKit::Display::Display):
* Source/WebKit/UIProcess/gtk/ScreenManagerGtk.cpp:
* Source/WebKit/UIProcess/gtk/SystemSettingsManagerGtk.cpp: Renamed from Source/WebKit/UIProcess/gtk/GtkSettingsManager.cpp.
(WebKit::SystemSettingsManager::themeName const):
(WebKit::SystemSettingsManager::darkMode const):
(WebKit::SystemSettingsManager::fontName const):
(WebKit::SystemSettingsManager::xftAntialias const):
(WebKit::SystemSettingsManager::xftHinting const):
(WebKit::SystemSettingsManager::xftHintStyle const):
(WebKit::SystemSettingsManager::xftRGBA const):
(WebKit::SystemSettingsManager::xftDPI const):
(WebKit::SystemSettingsManager::cursorBlink const):
(WebKit::SystemSettingsManager::cursorBlinkTime const):
(WebKit::SystemSettingsManager::primaryButtonWarpsSlider const):
(WebKit::SystemSettingsManager::overlayScrolling const):
(WebKit::SystemSettingsManager::enableAnimations const):
(WebKit::SystemSettingsManager::SystemSettingsManager):
* Source/WebKit/WebProcess/glib/SystemSettingsProxy.cpp: Renamed from Source/WebKit/Shared/gtk/GtkSettingsState.h.
(WebKit::SystemSettingsProxy::initialize):
(WebKit::SystemSettingsProxy::SystemSettingsProxy):
(WebKit::SystemSettingsProxy::didChange):
* Source/WebKit/WebProcess/glib/SystemSettingsProxy.h: Renamed from Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.h.
* Source/WebKit/WebProcess/glib/SystemSettingsProxy.messages.in: Renamed from Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.messages.in.
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::platformInitializeWebProcess):
(WebKit::WebProcess::platformTerminate):
* Source/WebKit/WebProcess/gtk/GtkSettingsManagerProxy.cpp: Removed.

Canonical link: https://commits.webkit.org/284171@main
